File: applications/rtktotalnuclearvariationdenoising/rtktotalnuclearvariationdenoising.cxx
/*=========================================================================
 *
 *  Copyright RTK Consortium
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *         https://www.apache.org/licenses/LICENSE-2.0.txt
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 *
 *=========================================================================*/

#include "rtktotalnuclearvariationdenoising_ggo.h"
#include "rtkGgoFunctions.h"
#include "rtkConfiguration.h"

#include "rtkTotalNuclearVariationDenoisingBPDQImageFilter.h"


#include <itkImageFileReader.h>
#include <itkImageFileWriter.h>

int
main(int argc, char * argv[])
{
  GGO(rtktotalnuclearvariationdenoising, args_info);

  using OutputPixelType = float;
  constexpr unsigned int Dimension = 4;           // Number of dimensions of the input image
  constexpr unsigned int DimensionsProcessed = 3; // Number of dimensions along which the gradient is computed

  using OutputImageType = itk::Image<OutputPixelType, Dimension>;
  using GradientOutputImageType = itk::Image<itk::CovariantVector<OutputPixelType, DimensionsProcessed>, Dimension>;
  using TVDenoisingFilterType =
    rtk::TotalNuclearVariationDenoisingBPDQImageFilter<OutputImageType, GradientOutputImageType>;

  // Read input
  OutputImageType::Pointer input;
  TRY_AND_EXIT_ON_ITK_EXCEPTION(input = itk::ReadImage<OutputImageType>(args_info.input_arg))

  // Apply total nuclear variation denoising
  auto tv = TVDenoisingFilterType::New();
  tv->SetInput(input);
  tv->SetGamma(args_info.gamma_arg);
  tv->SetNumberOfIterations(args_info.niter_arg);

  // Write
  TRY_AND_EXIT_ON_ITK_EXCEPTION(itk::WriteImage(tv->GetOutput(), args_info.output_arg))

  return EXIT_SUCCESS;
}

File: applications/rtktotalnuclearvariationdenoising/rtktotalnuclearvariationdenoising.py
#!/usr/bin/env python
import argparse
import itk
from itk import RTK as rtk


def build_parser():
    parser = rtk.RTKArgumentParser(
        description="Performs total nuclear variation denoising of a 3D + channels image."
    )
    parser.add_argument(
        "--input", "-i", help="Input file name", type=str, required=True
    )
    parser.add_argument(
        "--output", "-o", help="Output file name", type=str, required=True
    )
    parser.add_argument(
        "--gamma",
        "-g",
        help="TV term's weighting parameter",
        type=float,
        default=1.0,
    )
    parser.add_argument(
        "--niter",
        "-n",
        help="Number of iterations",
        type=int,
        default=5,
    )

    return parser


def process(args_info: argparse.Namespace):
    OutputPixelType = itk.F
    Dimension = 4  # Number of dimensions of the input image
    DimensionsProcessed = 3  # Number of dimensions along which the gradient is computed

    OutputImageType = itk.Image[OutputPixelType, Dimension]
    GradientPixelType = itk.CovariantVector[OutputPixelType, DimensionsProcessed]
    GradientImageType = itk.Image[GradientPixelType, Dimension]

    # Read input
    input_image = itk.imread(args_info.input, pixel_type=OutputPixelType)

    # Apply total nuclear variation denoising
    tv = rtk.TotalNuclearVariationDenoisingBPDQImageFilter[
        OutputImageType,
        GradientImageType,
    ].New()
    tv.SetInput(input_image)
    tv.SetGamma(args_info.gamma)
    tv.SetNumberOfIterations(args_info.niter)

    # Write
    itk.imwrite(tv.GetOutput(), args_info.output)


def main(argv=None):
    parser = build_parser()
    args_info = parser.parse_args(argv)
    process(args_info)


if __name__ == "__main__":
    main()
